x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -208,47 +219,116 @@ static void getMemoryAndProcessors(
                 extraModelMemoryInBytes += jobMemory;
             } else {
                 logger.debug("dfa job [{}] assigned to [{}], memory required [{}]", jobId, task.getAssignment(), jobMemory);
 
-                modelMemoryBytesSum += jobMemory;
+                existingModelMemoryBytes += jobMemory;
                 perNodeModelMemoryInBytes.computeIfAbsent(task.getExecutorNode(), k -> new ArrayList<>())
                     .add(MlJobRequirements.of(jobMemory, 0));
             }
         }
 
         // trained models
+        int processorsNewlyAssignedToTrainedModelsAllocationsThisTime = 0;
         for (var modelAssignment : autoscalingContext.modelAssignments.entrySet()) {
             TrainedModelAssignment assignment = modelAssignment.getValue();
-            final int numberOfAllocations = assignment.getTaskParams().getNumberOfAllocations();
+            final int numberOfRequestedAllocations = assignment.getTaskParams().getNumberOfAllocations();
             final int numberOfThreadsPerAllocation = assignment.getTaskParams().getThreadsPerAllocation();
             final long estimatedMemoryUsage = assignment.getTaskParams().estimateMemoryUsageBytes();
+            final int numTargetAllocationsOnExistingNodes = assignment.getNodeRoutingTable()
+                .values()
+                .stream()
+                .mapToInt(RoutingInfo::getTargetAllocations)
+                .sum();
+            final int numMissingAllocations = numberOfRequestedAllocations - numTargetAllocationsOnExistingNodes;
+            final int numProcessorsNeeded = numMissingAllocations * numberOfThreadsPerAllocation;
 
             if (AssignmentState.STARTING.equals(assignment.getAssignmentState()) && assignment.getNodeRoutingTable().isEmpty()) {
-
+                // this state occurs when the assignment has just been added and has not been assigned to any nodes
                 logger.debug(
                     () -> format(
                         "trained model [%s] lacks assignment , memory required [%d]",
                         modelAssignment.getKey(),
                         estimatedMemoryUsage
                     )
                 );
-
-                extraSingleNodeModelMemoryInBytes = Math.max(extraSingleNodeModelMemoryInBytes, estimatedMemoryUsage);
+                extraPerNodeModelMemoryBytes = Math.max(extraPerNodeModelMemoryBytes, estimatedMemoryUsage);
                 extraModelMemoryInBytes += estimatedMemoryUsage;
 
                 // if not low priority, check processor requirements
-                if (Priority.LOW.equals(modelAssignment.getValue().getTaskParams().getPriority()) == false) {
+                if (Priority.LOW.equals(modelAssignment.getValue().getTaskParams().getPriority()) == false
+                    && numProcessorsNeeded > numberOfAvailableProcessors(
+                        autoscalingContext,
+                        allocatedProcessorsScale,
```

Comment:
I don't understand this logic. Why do you need to add this?

And to start with: what is `extraProcessors` and `extraSingleNodeProcessors`?

Reply:
`extraProcessors` is the number of additional processors ES is requesting. It should be equal to the number of allocations * the number of threads per allocation which were not able to be assigned to any nodes. `extraSingleNodeProcessors` is the minimum number of processors per node, and should be equal to the `numThreadsPerAllocation`

This particular addition to this if-statement is needed because if the number of processors needed is less than or equal to the number of processors available on current nodes, we shouldn't request any extra processors. We might still need to request extra memory which might trigger a scaleup.